### PR TITLE
[4.2] Refactor repeatable to use arrayed names

### PR DESCRIPTION
**Pull request by pxpm**
This is a part of #3897 

This refactors repeatable field to use array names instead of stripping the names. 

Key changes:
- Validation is now like validating arrays in laravel `'your_field.*.name' => 'required'`
- Is now mandatory to cast this field in the model, either `array` or `json` would both work.

### CODE REVIEW

File: src/resources/views/crud/fields/select_multiple.blade.php [hunk 1]
```diff
@@ -6,6 +6,7 @@
         $options = call_user_func($field['options'], $field['model']::query());
     }
     $field['allows_null'] = $field['allows_null'] ?? true;
+    $field['value'] = isset($field['value']) && is_array($field['value']) ? $options->whereIn((new $field['model'])->getKeyName(), $field['value']) : collect();
```
Comment (tabacitu) [inline src/resources/views/crud/fields/select_multiple.blade.php:9]: This didn't make the `select_multiple` work inside `repeatable` for me. The adding seemed to work, but then on `edit` the saved options no longer show up as selected.

Same for `select2_multiple`, where copy-pasted this fix, thinking it works.

Comment (tabacitu) [reply src/resources/views/crud/fields/select_multiple.blade.php:9]: Could you please explain what this fixes and why it's needed @pxpm ?

File: src/resources/views/crud/fields/repeatable.blade.php [hunk 2]
```diff
@@ -34,41 +27,76 @@
...
+    @foreach($field['fields'] as $subfield)
+      @php
+          $subfield = $crud->makeSureFieldHasNecessaryAttributes($subfield);
+          $fieldViewNamespace = $subfield['view_namespace'] ?? 'crud::fields';
+          $fieldViewPath = $fieldViewNamespace.'.'.$subfield['type'];
+      @endphp
+      @include($fieldViewPath, ['field' => $subfield])
+    @endforeach
+  @endif
+</div>
```
Comment (tabacitu) [inline src/resources/views/crud/fields/repeatable.blade.php:98]: @pxpm I see we iterate through the fields twice now, in the same blade file. And output the same HTML twice - once inside the form, once outside it (pushed to `before_scripts`). 

I'll take that as a red flag 🚩  Why is this needed? Previously we were just doing that once.

File: src/resources/views/crud/fields/relationship/fetch.blade.php [hunk 3]
```diff
@@ -127,49 +133,6 @@
...
-                resolve($return);
-                    },
-                    error: function (result) {
-                        reject(result);
-                    }
-                });
-            });
-        };
-    }
-
```
Comment (tabacitu) [inline src/resources/views/crud/fields/relationship/fetch.blade.php:172]: Hmm.... why is this no longer needed, I wonder?

**Review by tabacitu — CHANGES_REQUESTED** [review]
Good work on this @pxpm 🎉🎉🎉

Seems to _mostly_ work 😀 , but I have a few questions regarding the implementation - please see my inline comments here.

File: src/app/Http/Controllers/Operations/FetchOperation.php [hunk 4]
```diff
@@ -61,16 +61,6 @@ private function fetch($arg)
...
-        // FetchOperation is aware of an optional parameter 'keys' that will fetch you the entity/entities that match the provided keys
-        if (request()->has('keys')) {
-            $array_keys = request()->get('keys');
-            if (is_array($array_keys)) {
-                return $model_instance->whereIn($model_instance->getKeyName(), $array_keys)->get();
-            } else {
-                return $model_instance->where($model_instance->getKeyName(), $array_keys)->get();
-            }
-        }
-
```
[the remark this answers is not in the document]
Comment (pxpm) [reply src/app/Http/Controllers/Operations/FetchOperation.php:73]: My rationale behing the removal of this was:
- This was used here to help dealing with fields inside repeatable (we needed to manually fetch some entries, no longer needed).
- The same result can be achieved customizing the `FetchOperation` **'query'** parameter inside the fetch function:
```php
public function fetchUser() {
        return $this->fetch([
            'model' => User::class,
            'query' => function($model_instance) {
                if (request()->has('keys')) {
                    $array_keys = request()->get('keys');
                    if (is_array($array_keys)) {
                        return $model_instance->whereIn($model_instance->getKeyName(), $array_keys)->get();
                    } else {
                       return $model_instance->where($model_instance->getKeyName(), $array_keys)->get();
                   }
                }
            },
            'searchable_attributes' => []
        ]);
    }
```
If we don't use it internaly I think we should not be maintaining it. Indeed it's a BC, a copy paste one ... but still a BC. I don't expect much or any people at all using this "hidden" functionality. On the other side, it's not a hard to maintain bit of code, we can just leave it as it is.

Last word is yours, I hold no strong argument about neither scenarios.

Let me know,
Pedro

File: src/resources/views/crud/fields/select_multiple.blade.php (hunk 1 above)
Comment (pxpm) [reply src/resources/views/crud/fields/select_multiple.blade.php]: This is needed here because in that field we expect the options to be a collection of models when retrived from the pivot table in the database. Ex: `$entry->tags` return a collection of `tag model` from `monster_tag` table. In repeatable the data is saved as a single array of id's inside the json. `tags => [1,3,5]`, so we need to convert it back to the actual models getting it from the retrieved options. 

For me it breaks with `Call to a member function pluck() on array` when doing `$field['value']->pluck()` because `$field['value']` is an array of selected id's. 

But indeed the conditional is wrong, I was just testing this in `dummy` and didn't see this breaks the regular fields. I've just submitted a fixed version and fixed in `select2_multiple` too. 

https://recordit.co/eKaUao4dYb

File: src/resources/views/crud/fields/repeatable.blade.php (hunk 2 above)
Comment (pxpm) [reply src/resources/views/crud/fields/repeatable.blade.php:98]: Kind of , in the first iteration we use the field value `@foreach ($field['value'] as $key => $row)` to create the appropriate field for each value stored in the database. For each row in the value we create the fields in the page. If there is no value (create operation for example) that iteration does not happen!

The second iteration is `the default state row`, we iterate the fields in repeatable creating the `empty state form fields` that we push to the DOM when needed. 

When there is value (1st iteration) we set the field value accordingly like when they are defined as a regular crud field in the update operation.

File: src/resources/views/crud/fields/relationship/fetch.blade.php (hunk 3 above)
Comment (pxpm) [reply src/resources/views/crud/fields/relationship/fetch.blade.php:172]: This is only used in: (master here) 
image

But even there, it introduces an inconsistency with other fields (like fetch, select_from_ajax etc) so I submitted a commit to also remove it from `fetch_and_create`.

File: src/app/Http/Controllers/Operations/FetchOperation.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (tabacitu) [reply src/app/Http/Controllers/Operations/FetchOperation.php:73]: Ah right, we have `query`. Ok - great thinking! In this case I think it's ok to remove, but let's leave this open until we write the docs for this - your explanation above MUST be in the upgrade guide.

Great work.

File: src/resources/views/crud/fields/select_multiple.blade.php (hunk 1 above)
Comment (tabacitu) [reply src/resources/views/crud/fields/select_multiple.blade.php]: Awesome! I can confirm it now works for entries submitted after this PR. 

However... it still didn't work for me when I tested using OLD entries. Let me tell you the story 😀



I've:
- added an entry in Dummies on the `master` branch, with the standard Relationship > select_multiple and select2_multiple filled in;
- switched the branch to this PR, reloaded the Edit page for that entry... now the entries do not show up as selected, even though they are selected in the database:


Here's a side-by-side comparison of two entries (left added on `master`, right added on `repeatable-as-array`, both of them previewed on `repetable-as-array`):

Screenshot 2021-11-11 at 10 30 44

I've done this to test that old projects' existing entries will still "just work" after the update. From what I understand... they don't.

I did a `dump($this->crud->getCurrentEntry()->relationships)` to see if there are any differences between what is stored in the database, and what do you know, there are:

Screenshot 2021-11-11 at 10 30 25

I see two big differences:
- on `master` the name is `categories[]`, but on `repeatable-as-array` it's `categories`;
- on `master` it's a JSON, but on `repeatable-as-array` it's an array;

I've done the `dump()` inside `setup()`, so it's way before it reaches the actual field blade file. 

Here's what I think:
- (1) We cannot expect people to change their DB entries, to use a new version of the `repeatable` field. It needs to work with the exact same DB entries. And when we save a new entry, what gets saved to look and work the same.
- (2) I was surprised to see that the name on `master` is `categories[]`... that's so ugly! It makes it super-difficult to use that value in your project... Seems like a huge mistake on our end...
- (3) Please investigate this and tell me what you find and let's decide what we can do about this.

File: src/resources/views/crud/fields/relationship/fetch.blade.php (hunk 3 above)
Comment (tabacitu) [reply src/resources/views/crud/fields/relationship/fetch.blade.php:172]: Oh ok. So let me rephrase this:
- Previously, on `master`, to show the selected entry we made an AJAX request to get its info (in order to show the Name not the Id, for example);
- Now, on `repeatable-as-array`, we determine the selected entry using PHP, not JS, so no need to make that AJAX request;

Is this correct?

If so, what are the benefits to making this change? Didn't we decide to populate the initial value with JS for a reason? I think so... I can't remember the reason, but... I think we had a good reason, right?

File: src/resources/views/crud/fields/select2_from_ajax_multiple.blade.php [hunk 5]
```diff
@@ -94,28 +94,8 @@ function bpFieldInitSelect2FromAjaxMultipleElement(element) {
         var $allowClear = element.attr('data-column-nullable') == 'true' ? true : false;
         var $dependencies = JSON.parse(element.attr('data-dependencies'));
         var $ajaxDelay = element.attr('data-ajax-delay');
-        var $selectedOptions = typeof element.attr('data-selected-options') === 'string' ? JSON.parse(element.attr('data-selected-options')) : JSON.parse("[]");
```
Comment (tabacitu) [inline src/resources/views/crud/fields/select2_from_ajax_multiple.blade.php:97]: Same thing happens here - right? Instead of populating the selected options using JS, we now do it in PHP?

File: src/resources/views/crud/fields/repeatable.blade.php (hunk 2 above)
Comment (tabacitu) [reply src/resources/views/crud/fields/repeatable.blade.php:98]: Ok, so that means... 
- the first one is the actual HTML that the user sees;
- the second one is a hidden HTML that we use to add new rows;

Why do we need this code duplication now, and didn't need it before? Because we now populate default repeatable values using PHP instead of JS? Why do we do that anyway?

**Comment by tabacitu** [comment]
Hey @pxpm ,

I've just replied to each of your responses, thank you 🙏 Please read and reply.

To sum up, though - I have a bad feeling about this. I expected this PR to involve a change inside `repeatable.blade.php`,  but it looks like this has balooned into changes across all our complicated fields. Which makes me scared to merge this.

From what I understand, this all stems from the fact that inside selects we now populate the default value using PHP instead of JS. Is this correct? I ask you then - why make this change? Do we really need to? Didn't we decide to populate the values with JS for a reason? I think so... though I can't remember the reason specifically (probably to make them work inside `repeatable` I guess). 

If possible, I'd rather keep the old behaviour of populating the values with JS, so that the changes are as few as possible.

Please let me know.
Cheers!

**Review by tabacitu — CHANGES_REQUESTED** [review]

File: src/resources/views/crud/fields/repeatable.blade.php (hunk 2 above)
Comment (pxpm) [reply src/resources/views/crud/fields/repeatable.blade.php:98]: We do that explicitly to avoid **initing empty fields** that technically **are not empty but we set the value later with JS**. 
This makes fields that depend on field value to setup stuff work inside repeatable, I am looking at you images and uploads ... probably datetimepicker too. Ahh and I think toggle is also in this list since it works in this branch and not in the master. 

This way fields are added to the page like any regular crud form field, we just setup the rows, selectors, array names etc after with JS, but on rendering, they are regular crud fields, with value and everything they deserve.

File: src/resources/views/crud/fields/select_multiple.blade.php (hunk 1 above)
Comment (pxpm) [reply src/resources/views/crud/fields/select_multiple.blade.php]: Hey @tabacitu indeed 2) is the problem. 

To make it work I added the check for `name` or `name[]` in the data we store in the database and it worked out of the box.

I pushed the fix here https://github.com/Laravel-Backpack/CRUD/pull/3905/commits/01436726c8e2a8588eadec6978de2db979f8914d

File: src/resources/views/crud/fields/relationship/fetch.blade.php (hunk 3 above)
Comment (pxpm) [reply src/resources/views/crud/fields/relationship/fetch.blade.php:172]: @tabacitu two different things here.

A) get the selected options inside repeatable
B) fetch default entry when field does not allow null. 

This one is specifically to **B)** because old me thought it was a good idea to have the `fetch_and_create` and the `fetch` fields have the "similar behavior" as a normal html5 select element, that is, if it does NOT allow null the first entry is selected by default. That is not what happens in select2 selects, and since this is a select2 select it's and inconsistency so should be removed.

Related to **A)** yes, we were only adding the value as JS because of the repeatable, those changes were made in the field to support repeatable. Since we are making repeatable array with major benefits there is no sense in keeping supporting this, this would be regular select2 fields with desired functionalities `fetch` or `fetch_or_create` (that I am thinking in merging into a single `fetch_and_create` but later we talk about it).

File: src/resources/views/crud/fields/select2_from_ajax_multiple.blade.php (hunk 5 above)
Comment (pxpm) [reply src/resources/views/crud/fields/select2_from_ajax_multiple.blade.php:97]: Like I said in other reply, this was only coded to support repeatable without BC in 4.1. I think there is no need to keep doing that if we don't need it.

File: src/resources/views/crud/fields/select_multiple.blade.php (hunk 1 above)
Comment (tabacitu) [reply src/resources/views/crud/fields/select_multiple.blade.php]: Can confirm it's working now, thanks @pxpm 🎉🎉🎉 

---

But **I don't think we're done here**. There's still a HUGE breaking change here. Imagine this scenario:
- dev developed admin panel in Backpack 4.1, using repeatable that stores categories as `categories[]`; codes his entire front-end app based on the information that this `[]` is at the end when he gets and sets categories outside his admin panel;
- dev upgrades his admin panel to Backpack 4.2, uses `repeatable` that stores categories as `categories`; his front-end will need to be changed too, because now there's no `[]` at the end;

Don't get me wrong. I hate the `categories[]` - never knew we had it. I'd say it's a bug, even. 

But if we do something like this (make the field basically rewrite the repeatable input name from `categories[]` to `categories`), then **we also have to tell the developer about it, in the upgrade guide. We need to tell him that the field will change the data structure in his database. That wherever he's using that data, he should assume it can be either `categories[]` or `categories`.** 

This makes it much more difficult to upgrade... They don't only have to change the validation, but also every bit of code that's using something stored using the `repeatable`.

----

I never wanted this 4.2 upgrade to be difficult. On the contrary, I wanted it to be a chinch. Even for the `repeatable` validation my plan was to tell people "_you can rewrite it to something like `'your_field.*.name' => 'required'`, which is much cleaner, but if you're too lazy to do that you can just add this line in your old validation, to make it work inside the repeatable v2_":

```diff
    public function rules()
    {
        return [
            'name'        => 'required|min:5|max:255',
            'description' => 'required',
            'simple'      => function ($attribute, $value, $fail) {
+               $fieldGroups = is_array($value) ? $value : json_decode($value);

                // do not allow repeatable field to be empty
                if (count($fieldGroups) == 0) {
                    return $fail('The simple field group must have at least one item.');
                }

                // ALTERNATIVE:
                // allow repeatable field to be empty
                // if (count($fieldGroups) == 0) {
                //   return true;
                // }

                // SECOND-LEVEL REPEATABLE VALIDATION
                // run through each field group inside the repeatable field
                // and run a custom validation for it
                foreach ($fieldGroups as $key => $group) {
                    $fieldGroupValidator = Validator::make((array) $group, [
                        'text'  => 'required|min:10|max:255',
                        'email' => 'required|email',
                    ]);

                    if ($fieldGroupValidator->fails()) {
                        // return $fail('One of the entries in the '.$attribute.' group is invalid.');
                        // alternatively, you could just output the first error
                        return $fail($fieldGroupValidator->errors()->first());
                        // or you could use this to debug the errors
                            // dd($fieldGroupValidator->errors());
                    }
                }
            },
        ];
    }
```

---

So let me ask you this. **Is there any benefit from introducing this `category[]` to `category` renaming, other than the fact that it's ugly?**

(A) Do we _really_ need to do this or could we just as well keep using the ugly `category[]` to keep the breaking changes to a minimum? 

(B) Maybe we could a flag/attribute on the `repeatable` field, something like `remove_trailing_brackets`; if `true` we do what you did now and rewrite; if `false` we keep the old `category[]` behaviour;

Not sure about this... what do you think @pxpm ? Aren't you worried this is becoming a pain to upgrade? We need to walk this difficult line between not disturbing our old `repeatable` users, when upgrading... and satisfying new `repeatable` users who would like to use any fields inside it...

File: src/resources/views/crud/fields/repeatable.blade.php (hunk 2 above)
Comment (tabacitu) [reply src/resources/views/crud/fields/repeatable.blade.php:99]: Ok got it. Last time we couldn't output with default directly, now we can. But it's very not-DRY to have the same code twice in the same file.

Can we create a function inside this file... say `outputRepeatableElements()` that will do the output, and we can call that twice? It would also need to check the function doesn't already exist - in case `repeatable` is used twice in the same form.

That would clean up the blade file a bit, don't you think?

File: src/resources/views/crud/fields/select2_from_ajax_multiple.blade.php (hunk 5 above)
Comment (tabacitu) [reply src/resources/views/crud/fields/select2_from_ajax_multiple.blade.php:97]: Ok, I was skeptical at first but... I really can't think of a reason we can't go back to defining things in PHP now... I think you're right. Closing.

File: src/resources/views/crud/fields/repeatable.blade.php [hunk 6]
```diff
@@ -48,11 +49,11 @@
                         $fieldViewNamespace = $subfield['view_namespace'] ?? 'crud::fields';
                         $fieldViewPath = $fieldViewNamespace.'.'.$subfield['type'];
                         if(!is_array($subfield['name'])) {
-                            if(isset($row[$subfield['name']])) {
-                                $subfield['value'] = $row[$subfield['name']];
+                            // this is a fix for 4.1 repeatable names that when the field was multiple, saved the keys with `[]` in the end. Eg: `tags[]` instead of `tags`
+                            if(isset($row[$subfield['name']]) || isset($row[$subfield['name'].'[]'])) {
+                                $subfield['value'] = $row[$subfield['name']] ?? $row[$subfield['name'].'[]'];
```
Comment (tabacitu) [inline src/resources/views/crud/fields/repeatable.blade.php:54]: Yeah... I think we should revert this 😅

File: src/resources/views/crud/fields/repeatable.blade.php [hunk 7]
```diff
@@ -352,15 +373,43 @@ function setupElementCustomSelectors(container) {
...
-        // update the container current number of rows by the amount provided
-        function updateRepeatableRowCount(container, amount) {
+        function updateRepeatableContainerNamesIndexes(container) {
+            container.children().each(function(i, repeatable) {
+                var index = $(repeatable).attr('data-row-number')-1;
+                // updates the indexes in the array of repeatable inputs
+                $(repeatable).find('input, select, textarea').each(function(i, el) {
+                    if(typeof $(el).attr('data-row-number') !== 'undefined') {
+                        let field_name = $(el).attr('data-repeatable-input-name') ?? $(el).attr('name') ?? $(el).parent().find('input[data-repeatable-input-name]').first().attr('data-repeatable-input-name');
+                        let unprefixed_field_name = field_name.endsWith("[]") ? field_name.substring(0, field_name.length - 2) : field_name;
```
[the remark this answers is not in the document]
Comment (tabacitu) [reply src/resources/views/crud/fields/repeatable.blade.php:383]: Note from our meeting - scratch that. Let's keep removing the `[]` (support both `[]` and no brackets for now). But new entries or edited entries will remove brackets.

File: src/resources/views/crud/fields/select2_multiple.blade.php [hunk 8]
```diff
@@ -12,6 +12,10 @@
 
     $field['multiple'] = $field['multiple'] ?? true;
     $field['allows_null'] = $field['allows_null'] ?? $crud->model::isColumnNullable($field['name']);
+
+    if(isset($field['value']) && is_array($field['value'])) {
+         $field['value'] = $field['options']->whereIn((new $field['model'])->getKeyName(), $field['value']);
+    }
```
[the remark this answers is not in the document]
Comment (pxpm) [reply src/resources/views/crud/fields/select2_multiple.blade.php:18]: Indeed @tabacitu `options_ids_array` is used in `Select All` button.

**Comment by pxpm** [comment]
@tabacitu please have a look at https://github.com/Laravel-Backpack/CRUD/pull/3905/commits/e01bd9f00e050461ec00a643dd02a40b45b7f1c4

It's using the blade file instead of the function, I think it's cleaner this way, and also easier to override if developer need something custom inside their repeatable container.

File: src/resources/views/crud/fields/repeatable.blade.php [hunk 9]
```diff
@@ -1,74 +1,50 @@
 {{-- REPEATABLE FIELD TYPE --}}
```
[the remark this answers is not in the document]
Comment (tabacitu) [reply src/resources/views/crud/fields/repeatable.blade.php:2]: (looks like the rows of the table field get doubled upon save)

**Review by tabacitu — CHANGES_REQUESTED** [review]
Much better @pxpm , thank you! 🎉🎉🎉 

I've tested here-and-there and found a few issues. Outlined below. I think we're on the right path here but still need to test the CRAP out of this, to make sure we're not breaking any fields.

So please do a run-though and test all fields inside repeatable and come back with confirmation that they're all working (and quick explanation why if they're not). 

Here's a copy of the list from [our previous issue in 4.1](https://github.com/Laravel-Backpack/CRUD/pull/2266), which were working there, so that we don't forget any of them:
- [ ] address_algolia
- [ ] address_google
- [ ] browse
- [ ] browse_multiple
- [ ] base64_image
- [ ] checkbox
- [ ] checklist
- [ ] checklist_dependency - see #2735;
- [ ] ckeditor
- [ ] color
- [ ] color_picker
- [ ] custom_html
- [ ] date
- [ ] date_picker
- [ ] date_range
- [ ] datetime
- [ ] datetime_picker
- [ ] easymde
- [ ] email
- [ ] ~enum~ - doesn't make sense inside Repeatable; use select_from_array instead;
- [ ] hidden
- [ ] icon_picker
- [ ] image
- [ ] month
- [x] number
- [ ] page_or_link -  see #2736;
- [x] password
- [x] radio
- [x] range
- [x] relationship
- [x] select (1-n relationship)
- [x] select_grouped
- [x] select2 (1-n relationship)
- [x] select_multiple (n-n relationship)
- [x] select2_multiple (n-n relationship)
- [x] select2_nested
- [x] select2_grouped
- [x] select_and_order
- [x] select_from_array
- [x] select2_from_array
- [x] select2_from_ajax
- [x] select2_from_ajax_multiple
- [x] simplemde
- [x] summernote
- [x] table
- [x] text
- [x] textarea
- [x] time
- [x] tinymce
- [ ] ~upload~ - currently impossible; see #2745
- [ ] ~upload_multiple~ - currently impossible; see #2745
- [x] url
- [x] video
- [x] view
- [x] week
- [x] wysiwyg

After you do that @pxpm , I'll do the same thing and test every one of them all over again and confirm each one is working. After I'm done, Antonio will do the same thing.

I think that will give us enough confidence we're not breaking any fields here.

**Comment by pxpm** [comment]
- [ ] address_algolia
- [ ] address_google
- [ ] browse
- [ ] browse_multiple
- [ ] base64_image
- [ ] checkbox
- [ ] checklist
- [ ] checklist_dependency - see #2735;
- [ ] ckeditor
- [ ] color
- [ ] color_picker
- [ ] custom_html
- [ ] date
- [ ] date_picker
- [ ] date_range
- [ ] datetime
- [ ] datetime_picker
- [ ] easymde
- [ ] email
- [ ] ~enum~ - doesn't make sense inside Repeatable; use select_from_array instead;
- [ ] hidden
- [ ] icon_picker
- [ ] image
- [ ] month
- [ ] number
- [ ] page_or_link -  see #2736;
- [ ] password
- [ ] radio
- [ ] range
- [ ] relationship
- [ ] select (1-n relationship)
- [ ] select_grouped
- [ ] select2 (1-n relationship)
- [ ] select_multiple (n-n relationship)
- [ ] select2_multiple (n-n relationship)
- [ ] select2_nested
- [ ] select2_grouped
- [ ] select_and_order
- [ ] select_from_array
- [ ] select2_from_array
- [ ] select2_from_ajax
- [ ] select2_from_ajax_multiple
- [ ] simplemde
- [ ] summernote
- [ ] table
- [ ] text
- [ ] textarea
- [ ] time
- [ ] tinymce
- [ ] ~upload~ - currently impossible; see #2745
- [ ] ~upload_multiple~ - currently impossible; see #2745
- [ ] url
- [ ] video
- [ ] view
- [ ] week
- [ ] wysiwyg

File: src/resources/views/crud/fields/repeatable.blade.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (pxpm) [reply src/resources/views/crud/fields/repeatable.blade.php:2]: I think that was related with a bug I had in code, initializing fields twice. 

I tested table field in the last state of the branch, and it was working, can you give it another try ? 

Thanks,
Pedor

**Comment by pxpm** [comment]
ping @tabacitu all fields that were working on previous version are working on this version, some small fixes here and there. Hope that I didn't miss much errors. Covered most simple cases. Going to have a quickshot in `upload_multiple`, atleast if we can fix that empty row from popping up, the uploading part in repeatable should be pretty much the same as in the `upload` field.
I am giving a chance to google field too, because I think it migth work as algolia is now working too. 

Let me know if you find something. 

Best,
Pedro

**Comment by tabacitu** [comment]
Exceleeeent! 🎉🎉🎉 

Thanks for bearing with me @pxpm . I know it's been hell redoing this `repeatable`... remember when we thought "_it'll probably take a few days_"? 😅

I've just finished testing ALL the fields (except the tricky new ones - `image`, `upload` and `upload_multiple`) and I can confirm the ones below are working. I've just found two bugs, really 
- `ckeditor` was impossible to remove from page
- `address_google` doesn't work (inside or _outside_ repeatable, very strange)

But other then that... zero, it all "_just worked_"! Good job @pxpm !!!

I'll do another quick code review tomorrow, with fresh eyes, just in case I missed something in terms of implementation details. But in terms of features... it works 🎉🎉🎉 Everything that worked in 4.1 works in 4.2! 🎉🎉🎉 

I say let's merge this PR as it is now, since it already has "_feature parity_" with 4.1. Let's finally merge it, so we can consider it done. And then for `image`, `upload` and `upload_multiple` we'll have separate PR / PRs, talk there etc.

My findings:

- [x] address_algolia
- [ ] address_google - does NOT work for me, neither in repeatable nor outside it; working API Key on email;
- [x] browse
- [x] browse_multiple
- [x] base64_image
- [x] checkbox
- [x] checklist
- [ ] ~~checklist_dependency~~ - see #2735;
- [x] ckeditor - could not be removed from page, but I fixed it;
- [x] color
- [x] color_picker
- [x] custom_html
- [x] date
- [x] date_picker
- [x] date_range
- [x] datetime
- [x] datetime_picker
- [x] easymde
- [x] email
- [ ] ~~enum~~ - doesn't make sense inside Repeatable; use select_from_array instead;
- [x] hidden
- [x] icon_picker
- [ ] image - haven't tested;
- [x] month
- [x] number
- [ ] ~~page_or_link~~ - see #2736;
- [x] password
- [x] radio
- [x] range
- [x] relationship
- [x] select (1-n relationship)
- [x] select_grouped
- [x] select2 (1-n relationship)
- [x] select_multiple (n-n relationship)
- [x] select2_multiple (n-n relationship)
- [x] select2_nested
- [x] select2_grouped
- [x] select_and_order
- [x] select_from_array
- [x] select2_from_array
- [x] select2_from_ajax
- [x] select2_from_ajax_multiple
- [x] ~~simplemde~~ - was removed in 4.2; loads easymde instead;
- [x] summernote
- [x] table
- [x] text
- [x] textarea
- [x] time
- [x] tinymce
- [ ] upload - haven't tested;
- [ ] upload_multiple - haven't tested;
- [x] url
- [x] video
- [x] view - works if the loaded view works;
- [x] week
- [x] wysiwyg

**Comment by tabacitu** [comment]
🎉🎉🎉🎉 Whohoooooo 🎉🎉🎉🎉
It's finally MERGED! And the docs for it.

happy_dance_cap

Thank you for your patience @pxpm - this was a BIG one, so much back-and-forth, I know it's been difficult 🙏
